src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPath.cs

```diff
@@ -171,5 +161,22 @@ internal void AddRange(ODataPath oDataPath)
         {
             this.segments.AddRange(oDataPath.segments);
         }
+
+        /// <summary>
+        /// Checks if this path is equal to another path.
+        /// </summary>
+        /// <param name="other">The other path to compare it to</param>
+        /// <returns>True if the two paths are equal</returns>
+        /// <exception cref="System.ArgumentNullException">Throws if the input other is null.</exception>
+        internal bool Equals(ODataPath other)
+        {
+            ExceptionUtils.CheckArgumentNotNull(other, "other");
+            if (this.segments.Count != other.segments.Count)
+            {
+                return false;
+            }
+
+            return !this.segments.Where((t, i) => !t.Equals(other.segments[i])).Any();

```

Comment:
Since this is now an indexable list, maybe using a plain for loop would be efficient, skipping the boxing and iterator overhead. But I don't know whether this is on a hot path.

Reply:
```suggestion
            for (int i = 0; i < this.segments.Count; ++i)
            {
                if (!this.segments[i].Equals(other.segments[i]))
                {
                    return false;
                }
            }
            
            return true;
```

Reply:
```suggestion
            return !this.segments.SequenceEquals(other.segments);
```

Reply:
I did a quick benchmark and the manual evaluation is best. Surprisingly, `SequenceEquals` is worst:
```
|        Method |      Mean |     Error |    StdDev | Allocated |
|-------------- |----------:|----------:|----------:|----------:|
|        Manual |  3.607 us | 0.0025 us | 0.0021 us |         - |
| SequenceEqual | 30.819 us | 0.2881 us | 0.2554 us |      48 B |
|   WhereNotAny | 18.368 us | 0.1094 us | 0.0913 us |      68 B |
```